SS14.Launcher/LauncherMessaging.cs

```diff
@@ -0,0 +1,135 @@
+                // Command is newline-terminated.
+                byte[] commandEncoded = Encoding.UTF8.GetBytes(string.Join('\n', commands) + "\n");
+                client.Write(commandEncoded);
+            }
+            Console.WriteLine("Passed commands to primary launcher");
+            return true;
+        }
+        catch (Exception)
+        {
+            // Ok, so we're server.
+            Console.WriteLine("We are primary launcher (or primary launcher is out for lunch)");
+        }
+        // Try to create server
+        try
+        {
+            _pipeServer = new NamedPipeServerStream(actualPipeName);
+            new Thread(() =>
+            {
+                try
+                {
+                    // Note we can't just close the StreamReader per-connection.
+                    // It would close the underlying pipe server (breaking everything).
+                    var sr = new StreamReader(_pipeServer, Encoding.UTF8);
+                    while (true)
+                    {
+                        _pipeServer.WaitForConnection();
+                        try
+                        {
+                            while (!sr.EndOfStream)
+                            {
+                                CommandQueue.Enqueue(sr.ReadLine());
+                            }
+                        }
+                        catch (Exception e)
+                        {
+                            // Not much we can do here.
+                            Console.WriteLine("Pipe server: Unexpected end of stream.");
+                        }
+                        try
+                        {
+                            _pipeServer.Disconnect();
+                        }
+                        catch (Exception)
+                        {
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Thrown when the pipeServer gets disposed.
+                    // Done on purpose so main thread can terminate the pipe server.
+                    // Note that ObjectDisposedException is not necessarily thrown!!!
+                    Console.WriteLine($"Pipe server: Shutting down, cause {e}");
+                }
+            }).Start();
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine($"Pipe server: Could not be created: {e}");
+        }
+
+        if (sendAnyway)
+        {
+            foreach (var command in commands)
+            {
+                CommandQueue.Enqueue(command);
+            }
+        }
+        return false;
+    }
+
+    /// <summary>
+    /// Closes the pipe server remotely, which causes WaitForConnection to throw, which cleans up the pipe server thread.
+    /// This is important because otherwise the thread sticks around.
+    /// </summary>
+    public static void ShutdownPipeServer()
+    {
+        if (_pipeServer != null)
```

Comment:
Use cancellation tokens for this instead. Also maybe make sure the thread has exited cleanly?

Reply:
Done.